authentication/middleware.py

```diff
@@ -33,3 +40,35 @@ def process_exception(self, request, exception):
                     urlquote(message), backend_name
                 )
                 return redirect(url)
+
+
+class BlockedIPMiddleware(MiddlewareMixin):
+    """
+    Only allow GET/HEAD requests for blocked ips, unless exempt or a superuser
+    """
+
+    def process_view(self, request, callback, callback_args, callback_kwargs):
+        """
+        Blocks an individual request if: it is from a blocked ip range, routable, not a safe request
+        and not from a superuser (don't want admins accidentally locking themselves out).
+
+        Args:
+            request (django.http.request.Request): the request to inspect
+        """
+
+        if (
+            not getattr(callback, "blocked_ip_exempt", False)
+            and not request.user.is_superuser
+            and not request.path.startswith("/admin/")
+        ):
+            user_ip, is_routable = get_client_ip(request)
+
+            if user_ip is None or (
+                is_routable
+                and request.method not in SAFE_METHODS
+                and BlockedIPRange.objects.filter(
+                    Q(ip_start__lte=user_ip) & Q(ip_end__gte=user_ip)
+                ).count()
+                > 0
+            ):
+                raise PermissionDenied
```

Comment:
This should `return django.http.HttpResponseForbidden()` instead, because django isn't aware of `rest_framework` exceptions so this turns into a server 500 instead.